### ISSUE TITLE
Remove Guides and AMs as admins from existing policies

### ISSUE THREAD

**Issue by davidcardoza** [Expensify/App#43493]
## Background
- Project tracking issue: [Expensify/Expensify#401707](https://github.com/Expensify/Expensify/issues/401707)
- Some AMs have elected to remain as admins on workspaces for specific customer accounts. These exceptions are listed in the following document: [List of AM Exceptions](https://docs.google.com/spreadsheets/d/1BPkkqyxRHi98b2qeY6Gam3S8ZT4G8PCUztS4j2YKKvs/edit#gid=0). We need to ensure these AMs are not removed as admins on their workspaces.
- A list of all existing Guide email accounts can be found [here](https://docs.google.com/spreadsheets/d/1emyRdwXWjxDhfEzgO_i51EaOsMkNb8wUnClnh5FQJTA/edit?gid=1172199479#gid=1172199479), and a list of all AM accounts can be found [here](https://docs.google.com/spreadsheets/d/1BPkkqyxRHi98b2qeY6Gam3S8ZT4G8PCUztS4j2YKKvs/edit?gid=0#gid=0).

## Tasks
1. **Review the List of AM Exceptions:**
   - Verify the AM exceptions listed in the provided document.
   - Ensure these domains and AMs are excluded from the removal process.

2. **Run the Script:**
   - Use the script from [Web-Expensify#42227](https://github.com/Expensify/Web-Expensify/pull/42227) to remove Guides and AMs from policies.
   - Check that the script does not remove AMs listed in the exceptions document.

3. **Ensure Continuation in #admins Room:**
   - Verify that Guides and AMs remain included in the #admins rooms in NewDot after the script has been run

## Non-engineering tasks
- Test the change once live with Guides and AMs

**Comment by tgolen** [comment]
Note for myself:

puneet
  7 minutes ago
[@tgolen](https://expensify.slack.com/team/U03TC9WA9)
 I think what we could do for your issue perhaps is just augment Ionatan's script with an optional flag of like addToAdminsRoom which would add the Guide/AM back to the admins room after removing them from the policy

**Comment by tgolen** [comment]
# Daily Update
- I have created a draft [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/42338) to allow some users to optionally be re-added to the #admin room after the policy is unshared with them
- I am currently trying to test this out locally to ensure it works first (I am running into an insufficient permissions issue, so I think I need to update Auth)
- @davidcardoza I would like to get a CSV of accounts to run with this bulk tool. I believe you can use the lists you posted to create the CSV. It needs to have three columns:
  - `email` of the person to remove
  - `policyID` of the policy to remove them from
  - `keepInAdminRoom` (value can be either "y", "n", or blank)

### Next Steps
- @tgolen once I have it tested and ensure it's working locally, I'll open the PR for full reviews.
- @iwiznia Give the draft PR a look to see if you spot any issues with what I've added
- merge and deploy the PR
- @davidcardoza construct the CSV file and deliver it to @tgolen 
- run the necessary removal script

#### ETA
- PR deployed by Thursday, June 13
- Script ran by Friday, June 14

**Comment by tgolen** [comment]
@puneetlath I've assigned this to you since you are taking it over and figuring out plan of attack.

**Comment by puneetlath** [comment]
Ok, so here's what I think we need to do. 

**Step 1: Remove Guides/AMs from all policies via script**
- To do this, we will use the BulkUnsharePolicies.php script
- We will pass it the BulkUnshare.csv
- This has about 63,000 policies, so with a 1 second sleep will take ~18 hours
- Removing them from the policy will also remove them from the #admins room

**Step 2: Add Guides/AMs back to #admins rooms via BulkCQ**
- Now we will add the Guides/AMs back to the admins rooms using the BulkCQ.py script
- We will pass it BulkCQ.csv
- We will use the query below

```sql
INSERT INTO sharedReports (
    accountID,
    permissions,
    reportID ) 
SELECT {0}, 'read, write', JSON_EXTRACT(value, '$.chatReportIDAdmins') as reportID
FROM nameValuePairs
WHERE name = '{1}'
ON CONFLICT (accountID, reportID) DO UPDATE SET permissions = excluded.permissions;
```

**Comment by iwiznia** [comment]
Plan looks good. TIL about the excluded syntax.

For the bulkCQ you can run probably something like 100 at a time

**Comment by cead22** [comment]
👍 Gonna first try this with 100 items, and then do the rest

**Comment by cead22** [comment]
That worked fine, so we're going to do the rest of the AMs, which are about 12K
[AMBulkCQ.csv.gpg.zip](https://github.com/user-attachments/files/15841069/AMBulkCQ.csv.gpg.zip)

**Comment by puneetlath** [comment]
Updated query to ignore conflicting lines

```sql
INSERT OR IGNORE INTO sharedReports (
    accountID,
    permissions,
    reportID ) 
SELECT {0}, 'read, write', JSON_EXTRACT(value, '$.chatReportIDAdmins') as reportID
FROM nameValuePairs
WHERE name = '{1}'
```

**Comment by cead22** [comment]
👍 to bulkCQ, we're going to use this file with 2600 entries
[AMRedoBulkCQ.csv.gpg.zip](https://github.com/user-attachments/files/15841350/AMRedoBulkCQ.csv.gpg.zip)

**Comment by puneetlath** [comment]
Ok, so far we have: 
1. Ran the BulkUnsharePolicies.php script for all AM accounts
2. Ran the CQ above to add all AMs back to #admins rooms
3. Ran the BulkUnsharePolicies.php script for all Guides

Next we need to re-add the Guides to the #admins rooms. To do that we will run BulkCQ with this query in batches of 100. 

```sql
INSERT OR IGNORE INTO sharedReports (
    accountID,
    permissions,
    reportID ) 
SELECT {0}, 'read, write', JSON_EXTRACT(value, '$.chatReportIDAdmins') as reportID
FROM nameValuePairs
WHERE name = '{1}'
```

It should take about 10 min to run.

**Comment by cead22** [comment]
👍 to bulk CQ, we'll run it with this file with 1s between batches [GuidesBulkCQ.csv.gpg.zip](https://github.com/user-attachments/files/15844131/GuidesBulkCQ.csv.gpg.zip)

**Comment by cead22** [comment]
We ran the bulkCQ and got 509 `200 OK` responses :tada

**Comment by cead22** [comment]
Waiting for an update on the state of things here https://expensify.slack.com/archives/C0745PP38PR/p1718643604728389?thread_ts=1718402793.809139&cid=C0745PP38PR

**Comment by cead22** [comment]
This is done and it looks like it did the trick, so closing

**Comment by aldo-expensify** [comment]
@cead22 could you help me with checking if `ricardo.lamothe@partnerhero.com` is in file shared here https://github.com/Expensify/App/issues/43493#issuecomment-2168802428 ? I can't open it because it was not encrypted for me.

I'm trying to see if that guide was missed somehow in the migration because it ended up causing this: https://github.com/Expensify/Expensify/issues/479487

**Comment by cead22** [comment]
I searched for accountID 12928064 in that file, but didn't find it. The file looks like this
```
7976487,expensify_policy78D8E408AF993F8B
7976487,expensify_policy35744D329177389F
7976487,expensify_policy042A2F642E2F28D8
7976487,expensify_policy0E68F2467EF600FB
7976487,expensify_policy346C95BC1336AB0D
7976487,expensify_policy389E2029996C6FD0
7976487,expensify_policyA0D933D0550D7F64
7976487,expensify_policyCC5237311C9AD668
7976487,expensify_policyEBFF87E0B1E6D3AF
7976487,expensify_policy13E720D407506D7A
```

```
carlos@db2.lax:~$ getacct.sh ricardo.lamothe@partnerhero.com
carlos@db2.lax:~$ getacct.sh ricardo.lamothe@team.expensify.com
12928064
```

**Comment by aldo-expensify** [comment]
Thanks for checking!

**Comment by aldo-expensify** [comment]
From reading this GH issue, I understand we got the guides to migrate from this file: https://docs.google.com/spreadsheets/d/1emyRdwXWjxDhfEzgO_i51EaOsMkNb8wUnClnh5FQJTA/edit?gid=1172199479#gid=1172199479

correct?

I don't find ricardo.lamothe there... I wonder what does that mean? is that list incomplete? is ricardo.lamothe not working anymore as a guide, but still had thousands of policies shared with him?

**Comment by cead22** [comment]
How do we do that? Happy to help with bulkCQs

**Comment by aldo-expensify** [comment]
I got a bit more information here: https://expensify.slack.com/archives/C03TN0B2V/p1755190686583199?thread_ts=1755051146.345249&cid=C03TN0B2V

I'm understanding that these were not really Guides, they are SecondResponders. I don't know how they got to be assigned to so many policies, but maybe that is why they were not migrated (didn't remove them from the policies)

I think we can do something like this:

1. Query accounts under the domains `@team.expensify.com` and maybe other external responders, and count on how many policies are they
2. Grab the ones that have lots of policies and lets remove them from the policies they shouldn't be

**Comment by puneetlath** [comment]
@aldo-expensify would you able to take this one? I'm off the next two days and am mostly focus on migration.

**Comment by aldo-expensify** [comment]
Yes, but I'm not sure those steps are incomplete. I got [here](https://github.com/Expensify/Expensify/issues/540742#issuecomment-3228262760) the accounts we need to fix, but I haven't have time to start working on it.

If these accounts missed the migration, maybe we also need to update the polices' assignedGuide (need to check that), or maybe that is just fine and we just have to remove these from the policies.

**Comment by aldo-expensify** [comment]
I think we can just close it for now, https://github.com/Expensify/Expensify/issues/479487 was removed from `#stability` since it has not happened for a long time and we have more important to focus on.
